### PR TITLE
Update type signature for Array#zip

### PR DESCRIPTION
**Pull request by ilyailya**
The PR updates the block parameter type in the RBI and the return type in the intrinsic for the Array#zip to match the type in the docs
https://docs.ruby-lang.org/en/3.1/Array.html#method-i-zip

### Motivation

https://github.com/sorbet/sorbet/issues/8206

### Test plan

See included automated tests.

### CODE REVIEW

**Comment by ilyailya** [comment]
We have a policy of testing changes to Sorbet against Stripe's codebase before
merging them. I've kicked off a test run for the current PR.  When the build
finishes, I'll share with you whether or how it failed. Thanks!

Stripe employees can see the build results here:

→ https://go/builds/bui_R0sQwYqHkd4WJB
→ https://go/builds/bui_R0sQ1rf8MhIpka
→ https://go/builds/bui_R0sQMQsenoidk1

File: rbi/core/array.rbi [hunk 1]
```diff
@@ -2799,7 +2799,7 @@ class Array < Object
   sig do
     type_parameters(:U).params(
         arg: T::Enumerable[T.type_parameter(:U)],
-        blk: T.proc.params(x: Elem, y: T.nilable(T.type_parameter(:U))).void
+        blk: T.proc.params(x: [Elem, T.nilable(T.type_parameter(:U))]).void
```
Comment (jez) [inline rbi/core/array.rbi:2802]: What's going on here?

Comment (jez) [inline rbi/core/array.rbi:2802]: I don't think that this solves the actual problem either—if you do

```ruby
xs = T.let([1,2,3], T::Array[Integer])
res = xs.zip(xs, xs) do |zipped|
  T.reveal_type(zipped)
end
```

Now `zipped` shows as

```
[Integer, T.nilable(Integer)] (2-tuple)
```

despite the correct type being

```
[Integer, T.nilable(Integer), T.nilable(Integer)] (3-tuple)
```

I would recommend that we split up the change to change the result type to `nil` from the change attempting to fix the argument type yielded to the block.

Off the top of my head, I'm not aware of any current intrinsics which attempt to model this distinction.

The easiest way to work around this in a way that's at least correct if unsatisfying would be to change the RBI like this


```suggestion
        blk: T.proc.params(x: T::Array[T.any(Elem, T.nilable(T.type_parameter(:U)))]).void
```

Possibly with an overload for the case where there is exactly one positional argument to annotate that the block will be yielded exactly two values.

**Review by jez — CHANGES_REQUESTED** [review]
^

File: rbi/core/array.rbi (hunk 1 above)
Comment (jez) [reply rbi/core/array.rbi:2802]: realized the problem. see below.

File: core/types/calls.cc [hunk 2]
```diff
@@ -4107,13 +4107,23 @@ class Array_zip : public IntrinsicMethod {
...
                 return;
             }
         }
 
-        res.returnType = Types::arrayOf(gs, make_type<TupleType>(move(unwrappedElems)));
+        if (args.block != nullptr) {
+            res.returnType = Types::nilClass();
+
+            if (auto *blockAppliedType = cast_type<AppliedType>(res.main.blockPreType)) {
+                ENFORCE(blockAppliedType->targs.size() == 2);
```
Comment (jez) [inline core/types/calls.cc:4121]:
```suggestion
                ENFORCE(blockAppliedType->targs.size() == 2, "calls.cc out of date w.r.t. array.rbi");
```

**Review by jez — APPROVED** [review]
Great find

**Pull request merged**